core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala

```diff
@@ -927,12 +927,15 @@ class UserService(protected val userInfo: UserInfo, val dataSource: SlickDataSou
   }
 
   def updateIntersectionGroupMembers(groupsToIntersect: Set[GroupsToIntersect], dataAccess:DataAccess): ReadWriteAction[Iterable[RawlsGroupRef]] = {
-    val groups = groupsToIntersect.flatMap { case GroupsToIntersect(target, group1, group2) => Seq(group1, group2) }.toSet
+    val groups = groupsToIntersect.flatMap { _.groups }
     val flattenedGroupsAction = DBIO.sequence(groups.map(group => dataAccess.rawlsGroupQuery.flattenGroupMembership(group).map(group -> _)).toSeq).map(_.toMap)
 
     val intersectionMemberships = flattenedGroupsAction.map { flattenedGroups =>
-      groupsToIntersect.map { case GroupsToIntersect(target, group1, group2) =>
-        (target, flattenedGroups(group1) intersect flattenedGroups(group2))
+      groupsToIntersect.map { case GroupsToIntersect(target, sourceGroups) =>
+        val flattenedGroupsToIntersect = flattenedGroups.filterKeys(sourceGroups)
+        target -> flattenedGroupsToIntersect.tail.values.fold(flattenedGroupsToIntersect.head._2) { (group1, group2) =>
```

Comment:
reduce here again I think. Reduce is a special case of fold where the starting value is the first element and the result is of the same type as the collection elements.